scipy/optimize/optimize.py

```diff
@@ -2606,12 +2609,35 @@ def _scalarfunc(*params):
         grid = grid[0]
         xmin = xmin[0]
     if callable(finish):
-        vals = finish(func, xmin, args=args, full_output=1, disp=disp)
-        xmin = vals[0]
-        Jmin = vals[1]
-        if vals[-1] > 0:
+        # set up kwargs for `finish` function
+        finish_args = getargspec(finish).args
+        kwargs = dict()
+        if 'full_output' in finish_args:
+            kwargs['full_output'] = 1
+        if 'disp' in finish_args:
+            kwargs['disp'] = disp
+        elif 'options' in finish_args:
+            # pass 'disp' as `options`
+            # (e.g. if `finish` is `minimize`)
+            kwargs['options'] = {'disp': disp}
+
+        # run minimizer
+        res = finish(func, xmin, args=args, **kwargs)
+
+        if isinstance(res, OptimizeResult):
+            xmin = res.x
+            Jmin = res.fun
+            success = res.success
+        else:
+            xmin = res[0]
+            Jmin = res[1]
+            success = res[-1] == 0
```

Comment:
@dlax What are you referring to as "previously"? I don't think I changed that line compared to my previous pull request. And in the original version before my pull request it was "if vals[-1] > 0" then print that it didn't succeed. That value is the statuscode number and it takes values 0, 1, 2, ..., with 0 corresponding to successful termination and > 0 to some error. I didn't find any error messages corresponding to numbers < 0. If the value is < 0 then it probably means that the function `finish` didn't have `statuscode` as the last parameter in the return list. Please let me know if I'm missing something.